Generation of API key on pbench-server

siddardh-ra opened the pull request:
Generate a unique API key internally when the user presents a Keycloak access_token and encode the user identity in the API key. 

Reworked `active_token` table.

This API Key will be also be validated under `verify_auth` module .

POST `/api/v1/generate_key` call generates api_key and will be updated in the reworked `api_key` (`active_token`)

dbutenhof reviewed (changes requested):
I started to review this before reading all of the main tab, but I realize you're not rebased on top of Nikhil's merge and you've got a lot of conflicts and obsolete code that you need to clean up before this is ready for review!

webbnh commented:
> POST `/api/v1/generate_key` call generates api_key and will be updated in the reworked `api_key` (`active_token`)

This API doesn't seem very resource-centric.  In particular, I expect that there will be a dcesire for an authorized user to be able to fetch the values of existing keys.  And, there will certainly be a need to be able to revoke/remove an existing key.  So, we'll want a suite of methods:
```
POST /api/v1/access_key    ## Creation...should this be PUT?  I can never keep them straight....
GET /api/v1/access_key
DELETE /api/v1/access_key/<id>   ## Not sure what the "ID" should be...the key value is ungainly....
```

npalaska commented:
Yeah, all the openid-connect changes are on the main now, you should rebase this on top of the main and that should minimize the conflicting files.

webbnh commented on lib/pbench/test/unit/server/auth/test_auth.py:
```diff
@@ -638,3 +608,65 @@ def tio_exc(token: str) -> JSON:
+        app.logger = make_logger
+        with app.app_context():
+            current_app.secret_key = jwt_secret
+            key, ts = pbench_drb_api_key
```
`ts` doesn't seem to be referenced; just use `_` instead.

webbnh commented on lib/pbench/test/unit/server/auth/test_auth.py:
```diff
@@ -638,3 +608,65 @@ def tio_exc(token: str) -> JSON:
+            current_app.secret_key = jwt_secret
+            key, ts = pbench_drb_api_key
+            user = Auth.verify_auth(key)
+        assert str(user.id) == DRB_USER_ID
```
Isn't `user.id` already a string?  Do we need the type conversion?

webbnh commented on lib/pbench/test/unit/server/auth/test_auth.py:
```diff
@@ -638,3 +608,65 @@ def tio_exc(token: str) -> JSON:
+        with app.app_context():
+            monkeypatch.setattr(oidc_client, "token_introspect", tio_exc)
+
+        """Verify success path of verify_auth_internal"""
```
This looks like a docstring...is there supposed to be a separate function here?

dbutenhof commented on lib/pbench/server/api/__init__.py:
```diff
@@ -130,6 +131,12 @@ def register_endpoints(api: Api, app: Flask, config: PbenchServerConfig):
         endpoint="endpoints",
         resource_class_args=(config,),
     )
+    api.add_resource(
```
Our REST APIs are `noun[/verb]` form; we're effectively managing an API Key resource here, so `/key/generate` might work; however it's also reasonable that a `POST /key` would *create* a key resource, and that also makes sense to provide an explicit `DELETE` on the same endpoint... whereas `DELETE /generate_key` makes no sense.

(Arguably `/upload/{name}` should have been `PUT /datasets`, which would return the `resource_id` or even a `location: /api/v1/datasets/{resource_id}`, but let's at least not add another weird one.)

dbutenhof commented on server/lib/config/pbench-server-default.cfg:
```diff
@@ -103,6 +103,9 @@ realm = pbench-server
 client = pbench-client
 
+# API key expiration duration in minutes, can be overridden in the main config file, defaults to 1 year
+api_key_expiration_duration = 525600
```
`api_key_duration_minutes` would be a more useful name!

webbnh replied:
Or, let's just get rid of expiration....  😏

webbnh commented on lib/pbench/test/unit/server/auth/test_auth.py:
```diff
@@ -638,3 +608,65 @@ def tio_exc(token: str) -> JSON:
+        oidc_client = OpenIDClient.construct_oidc_client(config)
+        monkeypatch.setattr(Auth, "oidc_client", oidc_client)
+
+        def tio_exc(token: str) -> JSON:
+            raise OpenIDTokenInvalid()
+
+        app = Flask("test-verify-auth-oidc-offline-invalid")
+        app.logger = make_logger
+        with app.app_context():
+            monkeypatch.setattr(oidc_client, "token_introspect", tio_exc)
+
+        """Verify success path of verify_auth_internal"""
```
Ditto my previous comments (1, 2) on lines 625-633.

dbutenhof commented on lib/pbench/test/unit/server/test_generate_key.py:
```diff
@@ -0,0 +1,38 @@
+
+    def test_successful_api_key_generation(self, query_get_as, pbench_drb_token):
+        response = query_get_as(pbench_drb_token, HTTPStatus.OK)
+        assert response.json["username"] == "drb"
```
As least assert that we *got* an API token as well as the username!

webbnh replied:
> At least assert that we got an API token as well as the username!

Yes, please.  (And, it would be nice if we could somehow establish that the API key was "valid" (e.g., by mocking `generate_key.generate_key()` or `jwt.encode()`), but perhaps that could be deferred to functional testing.)

And, btw..._why_ do we return the username?  Shouldn't the requestor already know its own username?

dbutenhof commented on lib/pbench/test/unit/server/test_endpoint_configure.py:
```diff
@@ -93,6 +93,7 @@ def check_config(self, client, server_config, host, my_headers={}):
                     },
                 },
                 "endpoints": {"template": f"{uri}/endpoints", "params": {}},
+                "generate_key": {"template": f"{uri}/generate_key", "params": {}},
```
Putting this here for lack of a better place: you need to add your new API name to the `API` ENUM in `pbench.client`, the lack of which is what broke the functional test run.

dbutenhof reviewed (changes requested):
Great start, Siddardh, but sadly, if not unexpectedly, I have comments... 😆

webbnh commented on lib/pbench/test/unit/server/conftest.py:
```diff
@@ -825,6 +826,27 @@ def get_token_func(pbench_admin_token, server_config, rsa_keys):
 
 
+@pytest.fixture()
+def pbench_drb_api_key(client, server_config, create_drb_user):
+    """Valid api_key for the 'drb' user"""
+    key, created_timestamp = generate_api_key(
+        username="drb",
+        user=create_drb_user,
+    )
+    return key, created_timestamp
+
+
+@pytest.fixture()
+def pbench_drb_api_key_invalid(client, server_config, create_drb_user):
+    """Invalid api_key for the 'drb' user"""
+    key, created_timestamp = generate_api_key(
+        username="drb",
+        user=create_drb_user,
+        valid=False,
+    )
+    return key
```
It strikes me as odd that the second of these two functions, which seems to mirror the first, doesn't return two values like the first does.  (Perhaps the first shouldn't return its second value?)

webbnh commented on lib/pbench/test/unit/server/conftest.py:
```diff
@@ -971,3 +993,46 @@ def tarball(tmp_path):
+        username: username to include in the token payload
+        user: user attributes will be extracted from the user object to include
+            in the token payload.
+        valid: If True, the generated token will be valid for 10 mins.
+            If False, generated token would be invalid and expired
```
s/would/will/ (to make the `False` verbiage match the `True` verbiage)

webbnh commented on lib/pbench/test/unit/server/database/test_api_key.py:
```diff
@@ -0,0 +1,64 @@
+        monkeypatch.setattr(Database, "db_session", __class__.session)
+        Database.Base.config = server_config
+        yield monkeypatch
+
+    def test_construct(self, fake_db, create_drb_user):
+        """Test api_key constructor"""
```
Why do we use `fake_db` here instead of `db_session`, like we did in the subsequent tests?

webbnh commented on lib/pbench/test/unit/server/database/test_api_key.py:
```diff
@@ -0,0 +1,64 @@
+
+        key, created_timestamp = pbench_drb_api_key
+
+        key2 = APIKey.query(key)
+        assert key2.api_key == key
```
Don't we want to check anything else in the database row, like the user ID or the creation timestamp?

webbnh commented on lib/pbench/test/unit/server/test_generate_key.py:
```diff
@@ -0,0 +1,38 @@
+
+        return query_api
+
+    def test_unauthorized_access(self, query_get_as, pbench_drb_token_invalid):
+        response = query_get_as(pbench_drb_token_invalid, HTTPStatus.UNAUTHORIZED)
+        assert response.json == {
+            "message": "User provided access_token is invalid or expired token"
+        }
```
What are we actually testing here?  I think we're supposed to be testing `/generate_key`, but the message is referring to an _`access_token`_ not an API key.

Is `pbench_drb_token_invalid` the correct fixture?  (Shouldn't we be using `pbench_drb_api_key_invalid`?)

A similar comment applies to `test_successful_api_key_generation()`.

siddardh-ra replied:
> What are we actually testing here? I think we're supposed to be testing `/generate_key`, but the message is referring to an _`access_token`_, not an API key.

Here we are testing whether the user is an authorized user or not (i.e) If a `user` hits this API with invalid `access_token` means, then the user won't be able to create an `api_key`

>Is pbench_drb_token_invalid the correct fixture? (Shouldn't we be using pbench_drb_api_key_invalid?)

Yes because `pbench_drb_api_key_invalid` will return a created `api_key` which is `expired` .(But I agree that ,we can also use `pbench_drb_api_key_invalid` fixture as well for authenticating it)

webbnh replied:
OK, I understand now.  (I made the above comment at the end of a long review and I was basically "out of gas".)

So, yes, this module tests the `/generate_key` API using _access tokens_ which are either valid or invalid...that makes good sense.  (There should be no need to test using the `pbench_drb_api_key_*` fixtures, because the application or effect of those aren't specific to the `/generate_key` API.)

webbnh reviewed (changes requested):
> Great start, Siddardh, but sadly, if not unexpectedly, I have comments... 😆

Yeah, what Dave said.  😉

A few overriding themes emerge:
- I think the API "resource" name should be a noun rather than a verb.
- What do we really want in the API key token values?  (They are _not_ OIDC/Keycloak tokens, so they don't have to have the same contents, I think.)
- Do we really want them to auto-expire (I think we do _not_).
- Some attention needs to be paid to what we log and how (e.g., we have a special exception for dealing with internal server errors, which we should use, and we should avoid calling `abort()` in deeply-nested functions).
- We need to support users having more than one key.

And, I've got a whole bunch of medium, small, and nit-level items.

webbnh commented on lib/pbench/server/api/resources/generate_key.py:
```diff
@@ -0,0 +1,147 @@
+                ApiMethod.POST,
+                OperationCode.CREATE,
+                authorization=ApiAuthorizationType.NONE,
+            ),
```
Yes, an API key is a stand-in for an authorization/credential -- the creation of one is a security-related event which should definitely be audited.

And, yes, we should have a `DELETE` method which allows an authorized user (or Admin) to revoke an API key.

And, although practice seems to be split on this, I think we should allow authorized users to query and fetch existing keys.

npalaska replied:
I wonder if adding api_key to audit should be a part of this PR but if it's not too much work then yeah we can add it I guess.

webbnh replied:
It would likely be easier to get it right if we add the auditing at the same time that we add the stuff which is supposed to be audited.

Due to simple evolution, we had to retrofit auditing into our existing code; however, going forward, we should avoid that.  New code should come with tests and any required audit/log calls -- these should not be afterthoughts.

webbnh commented on lib/pbench/server/api/resources/generate_key.py:
```diff
@@ -0,0 +1,147 @@
+                        user.username,
+                        new_key,
+                    )
+                except IntegrityError:
+                    current_app.logger.warning("Duplicate api_key got created")
+                    raise APIAbort(
+                        HTTPStatus.CONFLICT,
+                        message="Duplicate api_key created",
+                    )
```
> Why would we generate a duplicate key?

Because we got the same request twice at basically the exact same time?  (It _could_ happen....)

> Is this a user error?

Nope; it's an internal "concern" which we need to respond to appropriately, possibly by just returning the existing key value (it's valid, it's "right", and the fact that it's a duplicate doesn't mean it won't work if the user hands it back to us...).

> If you want to restrict users to a single API key (possibly not a bad idea)

As I said elsewhere, I believe that restricting the user to a single key _is_ a bad idea:  if they have some sort of automated system, they'll want the ability to migrate from an old key to a new one as the key values are rotated, which means that, for the migration period, they will need two key values _both_ of which are valid.  (Once the migration is complete, they can revoke the old key and prepare for the next rotation, so they never _need_ more than two keys, but they do need more than one.)

> see `_decode` in various model classes... and it'd really be nice to refactor this up into the model base class so it can be shared, along with `TZDateTime`

Oh, Dog!  Please do not copy that code **_again_** (we have, what, _four_ copies of it, now??).  They _so_ need to be refactored!!

webbnh commented on lib/pbench/server/auth/auth.py:
```diff
@@ -178,33 +137,23 @@ def verify_internal_token(auth_token: str) -> TokenState:
+        pass
+    except jwt.ExpiredSignatureError:
+        try:
+            APIKey.delete(api_key)
```
Yes, please.  We don't need to audit their _use_, any more than we audit access token use, but we should audit the creation and revocation of API keys.

npalaska replied:
I concur with what Webb said.

dbutenhof commented on lib/pbench/server/api/resources/generate_key.py:
```diff
@@ -0,0 +1,147 @@
+                    user=user,
+                )
+                try:
+                    key = APIKey(
+                        api_key=new_key, created=current_utc, expiration=expiration
+                    )
+                    user.update(api_key=key)
```
Yeah, I think we want to support multiple keys for a user, which might be used (and selectively revoked) for separate purposes. E.g., our Jenkins connection to GitHub is authenticated by Peter's API key, and when looking at a list of your tokens it might be nice to see "this one was for the <foo> CI, this one was for cluster performance data, etc."

So maybe replace the expiration timestamp with a *string* value which could be specified as a JSON body (e.g., `{"name": "CI key"}`), which would be retrieved via `GET` along with the key, and displayed in the UI to help with management?

siddardh-ra replied:
Good idea. We can implement this one. I need some clarifications on this, whether these key value JSON attributes be stored only in the DB while saving `api_key`?

>I don't know if we want to include this enhancement in this PR, but I suppose it's pretty simple

I'll add these in the next commit if there are changes requested or else I will raise a new PR for this

webbnh replied:
I believe that Dave's suggestion is that the `POST` method accept a JSON object as a payload from which the Pbench Server will extract a string (e.g., by looking for a key, `"name"`), and it will store that string in the database row with the API key value and the `User` reference.  We can worry later about how this string might be retrieved or updated.

However, if what we're doing is enhancing an existing method to now accept a payload (which we cannot yet `GET` or `PUT`), perhaps that would be better to do in a subsequent PR.

dbutenhof commented on lib/pbench/server/api/resources/generate_key.py:
```diff
@@ -0,0 +1,147 @@
+                        message="Duplicate api_key created",
+                    )
+                except SQLAlchemyError as e:
+                    current_app.logger.error(
+                        "SQLAlchemy Exception while generating an api_key for the user {}",
+                        type(e),
+                    )
+                    raise APIAbort(
+                        HTTPStatus.INTERNAL_SERVER_ERROR, message="INTERNAL ERROR"
+                    )
```
Using `APIInternalError` is important, because it generates a UUID that links the returned trivial error message to the full journal message.

npalaska commented on lib/pbench/server/api/resources/generate_key.py:
```diff
@@ -0,0 +1,147 @@
+        current_app.logger.exception(
+            "Could not encode the JWT api_key for user: {}", user
+        )
+        abort(
```
Another thing, I am not sure if this `abort` will work because Flask catches exceptions raised by the application and handles them internally, instead of propagating them up the call stack. This means that if you call abort() inside an exception block, Flask may not see the exception raised by `abort` and may not be able to handle it properly (it would just return 500). In this case you are already rasing 500 so you don't need an abort call at all.

webbnh replied:
@npalaska, your point is excellent (and it's part of the basis of our policy of not calling `abort()` from subroutines), but I think you have this backwards:

> Flask catches exceptions raised by the application and handles them internally, instead of propagating them up the call stack.

I believe that the up-stack code _is_ Flask, and the concern is that one of our `except Exception` blocks will catch the exception raised by `abort()` and "do the wrong thing" with it.  (This is part of the reason why `except Exception` blocks are not a good idea....)

> you are already raising 500 so you don't need an abort call at all.

I think this code is catching a JWT exception and trying to report the error.  It's true that we could let the exception propagate out and Flask would turn it into a `500` for us, but we really don't want that:  instead, we want to raise an `APIInternalError` instance, because that does useful stuff like creating an error ID, logging it, and reporting it out to the user.

npalaska commented on lib/pbench/server/api/resources/generate_key.py:
```diff
@@ -0,0 +1,147 @@
+                        user.username,
+                        new_key,
+                    )
+                except IntegrityError:
+                    current_app.logger.warning("Duplicate api_key got created")
+                    raise APIAbort(
+                        HTTPStatus.CONFLICT,
+                        message="Duplicate api_key created",
+                    )
```
This should probably be 500 as it's an error on our side if we generate a duplicate key.  Right?

webbnh replied:
I'm [not convinced](https://github.com/distributed-system-analysis/pbench/pull/3368#discussion_r1159019901) that it's an error at all.

npalaska commented on lib/pbench/server/auth/auth.py:
```diff
@@ -155,21 +114,21 @@ def verify_auth(auth_token: str) -> Optional[User]:
-    INVALID = enum.auto()
-    EXPIRED = enum.auto()
-    VERIFIED = enum.auto()
+    Tries to validate the api_key as if it was generated by the Pbench server .
```
But _it is_ generated by the pbench server

webbnh replied:
😆 

If you're actually objecting, @npalaska, you should suggest verbiage which would be acceptable.  😉

npalaska commented on lib/pbench/server/auth/auth.py:
```diff
@@ -178,33 +137,23 @@ def verify_internal_token(auth_token: str) -> TokenState:
                 "verify_exp": True,
```
Yes, if we are using a different validation path then I guess we can omit the audience field check.

npalaska commented on lib/pbench/server/database/models/api_key.py:
```diff
@@ -0,0 +1,63 @@
+    """Model for storing the API key associated with a user."""
+
+    __tablename__ = "api_key"
+    id = Column(Integer, primary_key=True, autoincrement=True)
```
And each API key should be unique so that's already a primary key.

npalaska commented on lib/pbench/server/database/models/api_key.py:
```diff
@@ -0,0 +1,63 @@
+    api_key = Column(String(500), unique=True, nullable=False, index=True)
+    created = Column(DateTime, nullable=False)
+    expiration = Column(DateTime, nullable=False)
+    user_id = Column(
+        String,
+        ForeignKey("users.id", ondelete="CASCADE"),
+        nullable=False,
+    )
```
Here `ondelete` is on a `ForeignKey` which means if this foreign key is deleted, then delete all the associated api_keys. However, we don't have a `delete` functionality on the user anymore (at least not through an API or via flask application, but we can always log into the PostgreSQL server and delete the user if we ever have to, but we discussed this earlier there is no point in deleting a user from our cached user table.) so yeah I guess this `ondelete` doesn't really mean much for this table.

webbnh replied:
> I guess this `ondelete` doesn't really mean much for this table.

Then we should remove it....

npalaska reviewed (commented):
Okay, I glanced over the changes but there are already a _lot_ of comments so I'll pass until you push another commit.

webbnh commented on lib/pbench/server/api/resources/generate_key.py:
```diff
@@ -0,0 +1,147 @@
+    user_obj = user.get_json()
+    current_utc = datetime.now(timezone.utc)
+    expiration = current_utc + time_delta
+    payload = {
+        "iat": current_utc,
+        "exp": expiration,
+        "user_id": user_obj["id"],
+        "username": user_obj["username"],
+        "audience": Auth.oidc_client.client_id,
+    }
+    try:
+        generated_api_key = jwt.encode(
+            payload, current_app.secret_key, algorithm=Auth._TOKEN_ALG_INT
+        )
```
But, in the proposed code, `verify_auth()` (via the subroutines that it calls) verifies API key tokens [independently](https://github.com/distributed-system-analysis/pbench/blob/c6be01a31d0a2d1da955179d2d301ac536871b97/lib/pbench/server/auth/auth.py#L178) of access tokens, so they don't need the same format.

npalaska replied:
Yep, I didn't look at all the files down before I made a comment.

webbnh replied:
Y'know...until you submit your review, you can go back and edit or delete the ("pending") comments which are part of the review....  🤓

dbutenhof commented on lib/pbench/server/database/models/api_key.py:
```diff
@@ -0,0 +1,103 @@
+        return repr(self.message)
+
+
+class APIKeys(Database.Base):
```
In all the other DB classes, the class name is *singular*; an instance represents a single "API key", not multiple "API keys". Only the table name is plural!

dbutenhof commented on lib/pbench/server/api/resources/api_key.py:
```diff
@@ -0,0 +1,104 @@
+)
+
+
+class APIKey(ApiBase):
```
Technically, having both `pbench.server.database.models.APIKey` and `pbench.server.api.resources.APIKey` isn't a problem for Python or necessarily for us, and since we generally have only one reference outside the class to the API implementation name, the risk of confusion is small.

Maybe something like `APIKeyManage`? 🍤

dbutenhof commented on lib/pbench/server/api/resources/api_key.py:
```diff
@@ -0,0 +1,104 @@
+                user=user,
+            )
+            key.add()
+        except IntegrityError:
```
You actually shouldn't ever be able to get this, because `add` catches any `Exception` and raises `APIKeyError`. But as I said before, when SQLAlchemy raises an `IntegrityError` it could mean either that you've violated your primary key uniqueness constraint (duplicate) *or* that you've violated one of the column `nullable = False` constraints, and there's no way to tell without the `_decode` logic we've already established has been duplicated too many times.

Still, without that, we shouldn't be blindly claiming "duplicate api key"; and since you're doing this as an internal server error anyway just *say* there was a database error.

Note that a `raise [...] from` both here and on the `APIKeyError` in `add` would make sure that we have the whole stack to analyze when it happens.

Philosophically, though, a "duplicate" means the user tried to generate two API keys closely enough together that the timestamps don't differ to the resolution we're using, and calling that an "internal server error" seems really unfair. (Both to us and to the user.)

dbutenhof commented on lib/pbench/server/api/resources/api_key.py:
```diff
@@ -0,0 +1,104 @@
+            key.add()
+        except IntegrityError:
+            raise APIInternalError("Duplicate api_key exists in the DB")
+        except SQLAlchemyError:
```
This also can't happen, again because `add` is converting any SQLAlchemy exception into `APIKeyError`.

dbutenhof commented on lib/pbench/server/database/models/api_key.py:
```diff
@@ -0,0 +1,103 @@
+            Database.db_session.commit()
+        except Exception as e:
+            Database.db_session.rollback()
+            self.logger.error("Can't add {} to DB: {}", str(self), str(e))
```
Interestingly, there's no `__str__` method defined, so the `str(self)` is not going to generate anything useful.

webbnh commented on lib/pbench/test/unit/server/auth/test_auth.py:
```diff
@@ -433,7 +433,7 @@ class TestAuthModule:
     encode_auth_token() since they are slated for removal and are covered well
     by other parts of the unit testing for the server code.
 
-    It does, however, verify the verify_auth_internal() method because we need
+    It does, however, verify the verify_auth() method because we need
     it to function properly until the use of an internal user is removed.
```
🤔  At this point, haven't we removed the use of an "internal user"?  And, isn't it clear by now that we're not intending to remove `get_current_user_id()`?  And, doesn't this PR remove `encode_auth_token()`?.

I think we need to touch up lines 432-437.

webbnh commented on lib/pbench/test/unit/server/auth/test_auth.py:
```diff
@@ -630,14 +630,13 @@ def tio_exc(token: str) -> JSON:
 
-        """Verify success path of verify_auth_internal"""
-        app = Flask("test-verify-auth-internal")
+        """Verify success path of verify_auth_api_key"""
+        app = Flask("test-verify-auth")
```
Line 633 looks like it should just be removed (we already have a better docstring at line 615).

The function argument in line 634 seems like it should be `test-verify-auth-api-key` to match the function name at line 612.  (And, line 628 should be removed, along with most if not all of the code in lines 616-632.)

webbnh commented on lib/pbench/test/unit/server/conftest.py:
```diff
@@ -1033,6 +1031,6 @@ def generate_api_key(
         "audience": Auth.oidc_client.client_id,
```
This token still includes the expiration time, which does nothing, now, right?

webbnh commented on lib/pbench/server/database/models/api_key.py:
To match the other files in the `models` directory, the name of this file should be plural.

webbnh commented on lib/pbench/test/unit/server/test_generate_key.py:
This module should be renamed to match `lib/pbench/server/api/resources/api_key.py`.

webbnh commented on lib/pbench/test/unit/server/auth/test_auth.py:
```diff
@@ -494,62 +493,33 @@ def record_abort(code: int, message: str = ""):
         with app.app_context():
             current_app.secret_key = jwt_secret
             user = Auth.verify_auth(pbench_drb_token)
         assert str(user.id) == DRB_USER_ID
```
Admittedly, you didn't add or change this, but `user.id` is already a string, so we don't need the conversion, here.

webbnh reviewed (changes requested):
Sorry to be posting/reviewing twice, but GitHub got confused and posted my previous review before I was ready....

It looks like you've addressed most of the items from my reviews from last week.  However, you missed a couple of items:  there's some odd code in `test_verify_auth_api_key()` ([1](https://github.com/distributed-system-analysis/pbench/pull/3368#discussion_r1158954527), [2](https://github.com/distributed-system-analysis/pbench/pull/3368#discussion_r1158958370)) and `test_verify_auth_api_key_invalid()` ([3](https://github.com/distributed-system-analysis/pbench/pull/3368#discussion_r1158959760)).  And there is the stuff that I found in your updates.

Also, apparently, this PR now has a conflict in `lib/pbench/server/api/__init__.py`, which you'll need to address before Mr. Jenkins will touch it.

And, I found one more thing which I request you fix, below.

siddardh-ra commented on lib/pbench/server/database/models/audit.py:
```diff
@@ -80,8 +80,8 @@ class AuditType(enum.Enum):
 
-    """Operation on an API token. There's no meaningful id or name."""
-    TOKEN = enum.auto()
+    """Operation on an API key."""
+    API_KEY = enum.auto()
```
Since we don't have unique information to add in `object_name` for audit currently, I haven't added anything about `object_name`

siddardh-ra commented:
> However, you missed a couple of items: there's some odd code in `test_verify_auth_api_key()` ([1](https://github.com/distributed-system-analysis/pbench/pull/3368#discussion_r1158954527), [2](https://github.com/distributed-system-analysis/pbench/pull/3368#discussion_r1158958370)) and `test_verify_auth_api_key_invalid()` ([3](https://github.com/distributed-system-analysis/pbench/pull/3368#discussion_r1158959760)). And there is the stuff that I found in your updates.
Addressed all the review comments  :+1: 

> And, I found one more thing which I request you fix, below.
Done

dbutenhof reviewed (changes requested):
The most important part is dropping the redundant audit `CREATE`; see the explanation in the comment. I'm also really on the edge about requesting that you set the completion audit `attributes` to record the key just in case we need to find it... but it's a value that's awkward to display.

dbutenhof commented on lib/pbench/test/unit/server/database/test_api_keys.py:
Although I wasn't 100% consistent here (I wish I had been), *most* of the tests in this directory end with `_db.py`, which makes it easy to distinguish in a list (for example with `-v` Pytest output) from the API test. I think that would be a good convention to follow rather than using the identical name in both directories.

webbnh commented on lib/pbench/server/api/resources/api_key.py:
```diff
@@ -62,31 +60,22 @@ def _post(
-            raise APIInternalError("Error while adding api_key to the DB")
+            raise APIInternalError(e.message) from e
+        except Exception as e:
+            raise APIInternalError(e) from e
```
So, if we get a duplicate key...are we now reporting that as an internal error?

webbnh reviewed (approved):
This is starting to look great!  I think you've resolved all of my actual objections (although, I second Dave's, but I'll let him be the one to block the merge, now 😀), but I did find some more things which you should consider addressing.

dbutenhof commented:
> This is starting to look great! I think you've resolved all of my actual objections (although, I second Dave's, but I'll let him be the one to block the merge, now grinning), but I did find some more things which you should consider addressing.

It seems that Siddardh acknowledged and even "liked" several comments, but didn't make changes. We need to remove the duplicate and broken audit record. I'd really really like to see the audit log verified in the unit tests, plus adding (and verifying) the audit `attributes` to identify the key would be great. I really don't like some aspects of the exception handling because right now the `APIKeyError` doesn't actually mean much but somewhat hides the real cause.

And Webb's right that, if this handling is turning a duplicate key into an internal server error, that's likely to become unpleasant ...

siddardh-ra commented:
> It seems that Siddardh acknowledged and even "liked" several comments, but didn't make changes. 

I've updated the changes related to `Audit` and also asserted that in unit tests as well and will push those changes.

>And Webb's right that, if this handling is turning a duplicate key into an internal server error, that's likely to become unpleasant.

Yeah, this is not the right way to handle this scenario. Adding some changes to handle the SQL Related errors for a debugging

webbnh commented on lib/pbench/server/database/models/__init__.py:
```diff
@@ -50,3 +51,44 @@ def process_result_value(self, value, dialect):
+        """
+        raise NotImplementedError()
+
+    def _decode(self, exception: IntegrityError) -> Exception:
```
The signature here is problematic:  I think there should be only one parameter -- the exception to be decoded...and that might be either `self` or `exception`, depending on whether this is really an instance method or a static method.

Also, since you're invoking this method outside the class, I don't think it should be named with the leading underscore prefix.

webbnh commented on lib/pbench/server/database/models/api_keys.py:
```diff
@@ -32,15 +35,38 @@ class APIKey(Database.Base):
-            self.logger.error("Can't add api_key to DB: {}", str(e))
-            raise APIKeyError(f"Error adding api_key to db : {e}") from e
+            self.logger.error("Can't add key:{} to DB: {}", str(self), str(e))
+            raise commonDBException._decode(self, e) from e
```
This is problematic:  you're invoking an instance method using a class reference, and so you have to explicitly provide what is normally an implicit parameter...which should raise various red flags.  😁 

I think you should be invoking `_decode()` using an instance of (a subclass of) the `commonDBException` class; so line 69 should be something more like
```python
            if isinstance(e, commonDBException):
                raise e._decode() from e
            else:
                raise
```
or
```python
            de = e._decode() if isinstance(e, commonDBException) else e
            if de is e:
                raise
            else:
                raise de from e
```

webbnh commented on lib/pbench/server/database/models/__init__.py:
```diff
@@ -50,3 +51,44 @@ def process_result_value(self, value, dialect):
+        # Postgres engine returns (code, message) but sqlite3 engine only
+        # returns (message); so always take the last element.
+        cause = exception.orig.args[-1]
+        if "UNIQUE constraint" in cause:
+            return self.DuplicateValue(self, cause)
+        elif "NOT NULL constraint" in cause:
+            return self.NullKey(self, cause)
```
The invocations of `DuplicateValue()` and `NullKey()` don't match the signatures at lines 59 and 67.

webbnh commented on lib/pbench/server/database/models/api_keys.py:
```diff
@@ -32,15 +35,38 @@ class APIKey(Database.Base):
+    def __str__(self):
+        return f"key: {self.api_key}"
+
+    class DuplicateValue(commonDBException):
```
The parent class, `commonDBException` requires that subclasses define `DuplicateValue()` and `NullKey()` instance methods, and the `DuplicateValue` class here does not do that.  (Ditto for the `NullKey` class at line 51.)

webbnh reviewed (changes requested):
Most of these changes are good; however, your attempt to refactor the SQL error decoding was not successful.  I think we need to try a somewhat different approach.  (I would be happy to discuss this with you, if you like.)

webbnh reviewed (commented):
> I would be happy to discuss this with you, if you like.

I went ahead and captured my thoughts.

webbnh commented on lib/pbench/server/database/models/__init__.py:
```diff
@@ -50,3 +51,44 @@ def process_result_value(self, value, dialect):
+        """
+        # Postgres engine returns (code, message) but sqlite3 engine only
+        # returns (message); so always take the last element.
+        cause = exception.orig.args[-1]
+        if "UNIQUE constraint" in cause:
+            return self.DuplicateValue(self, cause)
+        elif "NOT NULL constraint" in cause:
+            return self.NullKey(self, cause)
+        return exception
```
So, the situation here is that the knowledge of how to determine _which_ exception to raise should/needs to be encapsulated in this module, but the _identity_ (i.e., type) of the exceptions to be returned needs to be provided by the caller.  There are two obvious ways to accomplish that.

The first is for the caller to provide them as arguments to this function.  (They can be provided as instances, or as class objects, or as callables (e.g., factory functions).)  Under this approach, this `decode()` function doesn't need to be associated with any particular class -- it just needs to receive the exception instance to be decoded and the exception "ids" to return.

The second is to do as you've started to do here, to embed this function in the parent class of the exception instance to be decoded.  In this way, `self` will actually refer to a subclass instance, which means that this function (in the parent class) can invoke member functions in the subclass via `self` references.  But, I think it means that every instance of `self` has to know how to construct _both_ of the possibly-to-be-returned exception instances...which is slightly weird...and, then things start to unravel.

We have a number of class hierarchies, each descending from `Exception` via a separate custom base class (one for Audit, one for Server Settings, one for Templates, and one for Users), and they all derive a "duplicate" error and a "null-key" error from their respective base classes...which makes perfectly good sense, up until we start duplicating code.  I _think_ the ideal solution to this is a multiple inheritance approach (i.e., we have something which is, e.g., both an "Audit error" **and** an "SQL Constraint error").  But, it's not obvious to me how that factoid really helps us.

So, I'm inclined to say, for the immediate term, just use the first approach I outlined above:  make this `decode()` into a stand-alone utility function and have the caller just give it the two things that it should give back.  That will get us past the code duplication problem, and we can contemplate a more elegant solution at our leisure.

dbutenhof replied:
I don't particularly like the idea of passing fully built exception object to be tossed conditionally; but it'd work, and it does avoid the need for consistent constructor semantics across the models. Then again, I'd prefer to have consistent DB exception format anyway, and I prefer my idea of passing class references to construct on the fly.

webbnh replied:
I concur.  I think the right solution is to pass in factory methods which produce the required object when it's needed.  (It could just be a lambda expression on the caller's side.)  I don't much like the idea of passing in the class objects because it requires the decode function to assume that it knows what the c'tor signature is (and it requires all of the classes to have a common c'tor interface).

webbnh reviewed (commented):
> I would be happy to discuss this with you, if you like.

I went ahead and captured my thoughts.

dbutenhof commented on lib/pbench/test/unit/server/test_api_key.py:
```diff
@@ -0,0 +1,93 @@
+        assert audit[1].user_id == "3"
+        assert audit[1].user_name == "drb"
+        assert audit[1].reason is None
+        assert audit[1].attributes["key"]
```
Is there anything more specific we can assert here? I remember Webb making a similar comment about the JSON response. But here's a minimal idea ...

`assert audit[1].attributes["key"] == response.json["api_key"]`

webbnh commented on lib/pbench/server/database/models/api_keys.py:
```diff
@@ -20,40 +23,60 @@ def __str__(self):
-            self.logger.error("Can't add {} to DB: {}", str(self), str(e))
-            raise APIKeyError("Error adding api_key to db")
+            self.logger.error("Can't add key:{} to DB: {}", str(self), str(e))
+            raise commonDBException._decode(self, e) from e
```
Note that there are one or two other issues here:  if we cannot decode the exception and raise our own "duplicate key" or "null key" error, then we probably _don't_ want to reraise the original exception here, since it won't be an `APIKeyError` (or `commonDBException`) derivative; instead we presumably want to do something similar to what you were doing before and wrap the SQL exception (or, at least, its message) in an `APIKeyError`.  (The other issue is that `APIKeyError` and `commonDBException` don't have a common parent class (until you get all the way up to `Exception`), which means that the caller cannot catch them both without either catching them separately or catching all `Exception`'s...which is disappointing; hopefully, your rework of `_decode()` will also rectify the object hierarchies.)

webbnh commented on lib/pbench/server/database/models/api_key.py:
```diff
@@ -1,38 +1,58 @@
+    __tablename__ = "api_keys"
+    api_key = Column(
+        String(500), primary_key=True, unique=True, nullable=False, index=True
     )
```
@dbutenhof, @npalaska, am I correct in this assertion here?  (You cannot have a `primary_key` which is nullable or non-unique, right?  And, the whole point of the primary_key is that it _is_ the **index**, so it is superfluous to ask for an index for it, right?)

webbnh commented on lib/pbench/server/database/models/__init__.py:
```diff
@@ -50,3 +51,44 @@ def process_result_value(self, value, dialect):
         return value
+
+
+class commonDBException(Exception):
```
If this class were the base class for each of the `AuditError`, `ServerSettingError`, `TemplateError`, and `UserError` classes, then "common" would be an appropriate part of the name, but I'm not convinced that that is the hierarchy that we would want....

dbutenhof replied:
We could put a `class BaseDbError(Exception)` here, and interpose that under all of the other model class errors, if we wanted to have a generic `except` target; however I'm not convinced that would be particularly useful anywhere.

dbutenhof commented on lib/pbench/server/api/resources/api_key.py:
```diff
@@ -0,0 +1,76 @@
+        try:
+            key = APIKey(api_key=new_key, user=user)
+            key.add()
+        except DuplicateApiKey:
+            pass
```
In `upload`, we also "pass" on a duplicate, but return `OK` rather than `CREATED` so (if they care) the client can distinguish. Perhaps we should do that here as well rather than reporting `CREATED` when we did't actually.

For example, if a functional test created an API key, tested it, and then deleted it, and we for some reason ran two functional test suites in parallel, the `OK` status could inform the "loser" in the race that *deleting* that API key would be a bad idea as someone else might depend on it. (Not a problem until we implement `DELETE`, of course, but worth thinking about.)

dbutenhof commented on lib/pbench/server/database/models/__init__.py:
```diff
@@ -1,6 +1,8 @@
 import datetime
+from pyclbr import Class
```
This is wrong. And while I thought I remembered using `type[Class]` in the past, I can't find any example and I have to suspect my memory was clouded. 😆 

Apparently the correct type hint is just `type`, as in `on_null: type`; and remove this import. Sorry. But the good news is that the change might even let your `decode_integrity_error` declaration fit on one line. (It knocks off 14 characters.)

dbutenhof commented on lib/pbench/server/database/models/api_keys.py:
```diff
@@ -0,0 +1,128 @@
+        self.cause = cause
+
+    def __str__(self) -> str:
+        return f"Missing required key: {self.cause}"
```
I keep telling myself "maybe this is OK", but every time I see it something says "it's a missing *value* not a missing *key*". Hmm. Then again, it's not necessarily "missing", as in `user_id = None` isn't "missing" but would still trigger the integrity error. I can also tell myself this shouldn't matter much as we're defaulting the `created` column in all cases, which defaults to the current time, we already would have failed creation if we somehow *generated* a null API key, and certainly if we didn't find a user ID from `Auth`... and even if it somehow happens, this only translates to a journal message we're not directly reporting to the client.

dbutenhof commented on lib/pbench/server/database/models/api_keys.py:
```diff
@@ -0,0 +1,128 @@
+            Database.db_session.commit()
+        except Exception as e:
+            Database.db_session.rollback()
+            self.logger.error("Can't add key:{} to DB: {}", str(self), str(e))
```
Using the `__str__` above, this is "Can't add key:key {key} to DB: {message}", which is ... awkward. Not a big deal as it's just in the journal, but if you touch this file again, you might consider making the `__str__` more explicit with something like "API key {self.api_key}" and dropping the redundant "key:" here.

dbutenhof commented on lib/pbench/server/database/models/api_keys.py:
```diff
@@ -0,0 +1,128 @@
+        Returns:
+            An APIKey object if found, otherwise None
+        """
+        # We currently only query api_key database with given api_key
```
This strikes me as a fairly useless comment. It isn't really clear, and is mostly redundant with the simple docstring.

webbnh commented on lib/pbench/server/database/models/api_keys.py:
```diff
@@ -0,0 +1,128 @@
+        except Exception as e:
+            Database.db_session.rollback()
+            self.logger.error("Can't add key:{} to DB: {}", str(self), str(e))
+            raise decode_integrity_error(
+                e, on_duplicate=DuplicateApiKey, on_null=NullKey
+            )
```
If we're going to raise a new exception here, then we should have a `from` clause on the `raise`.  Presumably, just `from e`, which, hopefully, will do the right thing in all the cases.

(I don't know whether `raise e from e` has the same semantics as a bare `raise` -- it would be handy if it did -- but, if it doesn't, it will at least have a better result than we would get omitting the `from`.)

webbnh replied:
My [previous concern](https://github.com/distributed-system-analysis/pbench/pull/3368#discussion_r1164667589) still applies, here:  if `decode_integrity_error()` returns the original exception, we probably don't want to (re)raise it here, since it's not a subclass of `APIKeyError`.  Instead, we should probably raise an instance of that, instead, with the stringization of the original exception.

webbnh reviewed (commented):
Yes, I'm with Dave:  we're getting really close.  But, there are a few items (although small, at this point) which need touching up.

In particular, the _call_ to `decode_integrity_error()` needs a little more work, and I recommend being a little more rigorous and concrete with its interface.

dbutenhof commented on lib/pbench/server/api/resources/api_key.py:
```diff
@@ -66,7 +66,9 @@ def _post(
-            pass
+            response = jsonify({"api_key": new_key})
+            response.status_code = HTTPStatus.OK
+            return response
```
This is also skipping the audit `attributes` setting ...

Although ... this makes me realize that there isn't really a clear way to record the "duplicate, alternate success" status, and I'm not sure what to do about that. Probably nothing, in the context of this PR. 😦

webbnh replied:
> This is also skipping the audit attributes setting ...

**Another** excellent reason for a common return point!

> I'm not sure what to do about that. Probably nothing, in the context of this PR. 😦

Yes, not in this PR....  😏

dbutenhof commented on lib/pbench/server/database/models/api_keys.py:
```diff
@@ -66,9 +66,13 @@ def add(self):
+            decode_exc = decode_integrity_error(
                 e, on_duplicate=DuplicateApiKey, on_null=NullKey
             )
+            if decode_exc == e:
```
This should probably be `if decode_exc is e` ... that is, it's not just `Object.__eq__(e)` (whatever logic the superclass `__eq__` chooses, which might not be much) but the exact same exception instance. Although in the context of this logic I'm not sure it makes a functional difference since the other paths aren't likely to return an equal or identical instance. 😆

dbutenhof commented on lib/pbench/server/database/models/__init__.py:
```diff
@@ -53,42 +54,24 @@ def process_result_value(self, value, dialect):
-class commonDBException(Exception):
-    """Generic base class for reporting errors ."""
+def decode_integrity_error(
+    exception: IntegrityError, on_null: type[Class], on_duplicate: type[Class]
```
I don't see how that distinction really matters here; either way we need a compatible signature, unless we take an earlier idea I threw out of relying on passing through `**kwargs` without any fixed arguments.

Still, `isinstance(<Classname>, Callable)` is True, and a factory method here would work just as well as a direct constructor, so I wouldn't argue even if I don't see much value.

dbutenhof reviewed (approved):
Just some commentary on recent changes and conversations, but I'm happy enough at this point to just get it in.

webbnh reviewed (approved):
This is good enough, now.  😀

It would be better with the `is`/`==` thing and the `decode_integrity_error()` signature fixed, and it would be nicer if `decode_integrity_error()` took an additional argument to construct the no-match return value, but I think we can merge this as it is, now (we can add the third factory to `decode_integrity_error()` when we convert the other three DB classes to use it).

dbutenhof commented:
Let's get this in, but a suggestion for a small task for @siddardh-ra ... we should add a *functional* test case that uses your `POST /api/v1/key` to generate an API key and validates that you can use it to see/update `tester` owned private datasets.

I hate the way our functional test system is getting forced into a single file in order to get test dependency ordering (because the `@pytest.mark.dependency` only works within a file); but it can be a separate *class*, and dependent on upload.
